src/Interpreters/CollectJoinOnKeysVisitor.cpp

```diff
@@ -71,32 +102,50 @@ void CollectJoinOnKeysMatcher::visit(const ASTFunction & func, const ASTPtr & as
             throw Exception("Function " + func.name + " takes two arguments, got '" + func.formatForErrorMessage() + "' instead",
                             ErrorCodes::SYNTAX_ERROR);
     }
-    else
-        throw Exception("Expected equality or inequality, got '" + queryToString(ast) + "'", ErrorCodes::INVALID_JOIN_ON_EXPRESSION);
 
     if (func.name == "equals")
     {
         ASTPtr left = func.arguments->children.at(0);
         ASTPtr right = func.arguments->children.at(1);
-        auto table_numbers = getTableNumbers(ast, left, right, data);
-        data.addJoinKeys(left, right, table_numbers);
+        auto table_numbers = getTableNumbers(left, right, data);
+        if (table_numbers.first == table_numbers.second)
+        {
+            if (table_numbers.first == JoinIdentifierPos::Unknown)
+                throw Exception("Ambiguous column in expression '" + queryToString(ast) + "' in JOIN ON section",
+                                ErrorCodes::AMBIGUOUS_COLUMN_NAME);
+            data.analyzed_join.addJoinCondition(ast, isLeftIdentifier(table_numbers.first));
+            return;
+
+        }
+        if (table_numbers.first != JoinIdentifierPos::NotApplicable && table_numbers.second != JoinIdentifierPos::NotApplicable)
+        {
+            data.addJoinKeys(left, right, table_numbers);
+            return;
+        }
     }
-    else if (inequality != ASOF::Inequality::None)
+
+    if (auto expr_from_table = getTableForIdentifiers(ast, false, data); expr_from_table != JoinIdentifierPos::Unknown)
     {
-        if (!data.is_asof)
-            throw Exception("JOIN ON inequalities are not supported. Unexpected '" + queryToString(ast) + "'",
-                            ErrorCodes::NOT_IMPLEMENTED);
+        data.analyzed_join.addJoinCondition(ast, isLeftIdentifier(expr_from_table));
```

Comment:
And this one doesn't allow functions other than `equals` to have identifiers from different sides (left/right tables). Also should add test on these functions.

Reply:
Yes, `... ON ... AND t1.x > t2.x` not supported yet, I'll add test. But it works in postrgess. I think I'll try to support it later in other PR.

Reply:
https://github.com/vdimir/ClickHouse/blob/de3198020cfa05275c9c63a16657fd2f18ec3d45/tests/queries/0_stateless/01881_join_on_conditions.sql#L52-L54